poller-service.py

```diff
@@ -0,0 +1,286 @@
+
+
+def getLock(lock):
+    global cursor
+    query = "SELECT GET_LOCK('{}', 0)".format(lock)
+    cursor.execute(query)
+    return cursor.fetchall()[0][0] == 1
+
+
+def releaseLock(lock):
+    global cursor
+    query = "SELECT RELEASE_LOCK('{}')".format(lock)
+    cursor.execute(query)
+    return cursor.fetchall()[0][0] == 1
+
+
+def sleep_until(timestamp):
+    now = datetime.now()
+    if timestamp > now:
+        sleeptime = (timestamp - now).seconds
+    else:
+        sleeptime = 0
+    time.sleep(sleeptime)
+
+poller_group = ('and poller_group IN({}) '
+                .format(str(config['distributed_poller_group'])) if 'distributed_poller_group' in config else '')
+
+# Add last_polled and last_polled_timetaken so we can sort by the time the last poll started, with the goal
+# of having each device complete a poll within the given time range.
+dev_query = ('SELECT device_id, status,                                          '
+             'DATE_ADD(                                                          '
+             '  DATE_SUB(                                                        '
+             '    last_polled,                                                   '
+             '    INTERVAL last_polled_timetaken SECOND                          '
+             '  ),                                                               '
+             '  INTERVAL {0} SECOND) AS next_poll,                               '
+             'DATE_ADD(                                                          '
+             '  DATE_SUB(                                                        '
+             '    last_discovered,                                               '
+             '    INTERVAL last_discovered_timetaken SECOND                      '
+             '  ),                                                               '
+             '  INTERVAL {1} SECOND) AS next_discovery                           '
+             'FROM devices WHERE                                                 '
+             'disabled = 0                                                       '
+             'AND IS_FREE_LOCK(CONCAT("polling.", device_id))                    '
+             'AND IS_FREE_LOCK(CONCAT("queued.", device_id))                     '
+             'AND ( last_poll_attempted < DATE_SUB(NOW(), INTERVAL {2} SECOND )  '
+             '  OR last_poll_attempted IS NULL )                                 '
+             '{3}                                                                '
+             'ORDER BY next_poll asc                                             ').format(poll_frequency,
+                                                                                           discover_frequency,
+                                                                                           down_retry,
+                                                                                           poller_group)
```

Comment:
Here, the string `poller_group` is substitude for {3}. So if `poller_group` contains additional paramaters for the query, they are included, if it is the empty string, it has no effect on the query.